utilities/src/main/java/ru/study21/jcsv/xxl/io/DefaultCSVReader.java

```diff
@@ -34,15 +36,15 @@ public CSVMeta meta() {
         return _meta;
     }
 
-    public List<String> nextRow() throws BrokenContentsException {
+    public CSVRow nextRow() throws BrokenContentsException {
         List<String> result = _record;
         try {
             _record = _recordSupplier.parse(_reader);
         } catch (IOException e) {
             // TODO: fix it
             throw new RuntimeException(e);
         }
-        return result;
+        return new CSVRow(result, _offset++);
```

Comment:
When I said "offset" I didn'n mean row number in terms of csv abstraction.

I meant position in file of first byte in record. 

Overall idea is following. 
We need to track for each row where to pick-up its contents when huge file is sorted.

For example, when sort is over and we start to write results let's assume that first row after sorting is row number 4245245 of original file. And we must know where to go in original file to read it's contents (which can be much longer then keys participated in sort - that's why we don't pass whole row via sorting).

So we  need to keep in memory byte offset for row #4245245  as well as each other row.

The question is how to know it.

There are 2 basic options:

1. Keep track manually. I don't like it because there is risk that some abstract BufferedReader will merge internally '\r' and '\n' or something like that. And idea of java.io is that many kinds of stuff can stand behind BufferedReader. It can be zipped file, network socket and so on. Any kind of position makes little sense here for us because we can't come back to it for data.

2. Ask file system. It is best way. But I don't want to make our CSV parsing logic less abstract.

To solve this contradiction lets introduce new class like "CSVSource" that would be record with 2 fields: BufferedReader and Optional<FileChannel>. If CSVSource is created on top of real file then optional FileChannel will keep value from FileInputStream.getChannel,  Otherwise it will be empty.

And inside nextRow we'll use real file position if there is real file and -1 otherwise.